p4-16/psa/PSA.mdk

```diff
@@ -2572,3 +2543,139 @@ Certainly if there are long durations of time when egress decides to
 drop all packets to an output port, that port will go idle.  The
 scheduling algorithm implementations are all built with a finite
 maximum packet scheduling rate.
+
+
+# Appendix: Packet ordering {#appendix-packet-ordering}
+
+This section describes _recommendations_ for PSA implementations on
+the order that packets are processed.  These are not requirements,
+since there are known implementation techniques, especially
+parallelism that can be taken advantage of in a variety of ways, that
+can lead to cheaper implementations if these recommendations are not
+followed.  We recommend that developers selecting P4 devices ask their
+designers about these issues, if those developers consider the issues
+important for their purposes.
+
+Recommendation 1: Packets that arrive on the same input port should
+begin ingress processing in the same relative order as they arrived.
+
+Recommendation 2: Packets that go out on the same output port should
+be transmitted on the port in the same relative order that they began
+egress processing.
+
+Recommendation 3: PRE unicast packets (i.e. those that follow the
+"enqueue one packet" path in the pseudocode of section
+[#sec-after-ingress]) that arrived from the same ingress port, did
+ingress processing once (i.e. were not resubmitted or recirculated),
+were sent to the PRE with the same `class_of_service` value, and
+destined for the same egress port, should begin egress processing in
+the same relative order as they began ingress processing.
+
+It is expected that some PSA implementations will implement the class
+of service mechanism by having a separate FIFO queue per class of
+service, and thus while unicast packets with the same ingress port,
+egress port, and class of service will pass through the system in FIFO
+order if they follow all recommendations above, unicast packets with
+the same ingress and egress port, but different classes of service,
+may be processed by the egress control block in a different order than
+they were processed by the ingress control block.
+
+If an implementation satisfies recommendations 1 through 3, then
+unicast traffic assigned to the same class of service will maintain
+its relative order through the device.
+
+Recommendation 4: PRE multicast packets (i.e. those that follow the
+"Make 0 or more copies" path in the pseudocode of section
+[#sec-after-ingress]) that arrived on the same ingress port, did
+ingress processing once, were sent to the PRE with the same pair of
+values (`class_of_service`, `multicast_group`), for copies destined to
+the same egress port, and with the same pair of values for
+(`egress_port`, `instance`), should begin egress processing in the
+same relative order as they began ingress processing.
```

Comment:
Aren't the copies "spawned" in the PRE? Therefore, they have not gone through ingress processing, so I suppose the relative order refers only to the source packet for multicast. If the copies are to be ordered, then they need to inherit the ordering from the source.

Reply:
I have tried rewording Recommendation 4 a little in order to make the original packet vs. copy distinction more explicitly.